File: crates/ferric-rules-ffi/tests/c/output_lifetime.c
/*
 * output_lifetime.c — C-ABI regression harness for FR-CABI-004 (issue #114).
 *
 * A borrowed output pointer belongs to one engine. Reading the same channel
 * from a second engine must not invalidate the first engine's pointer. The
 * copy API must report its exact size and any truncation, and repeated engine
 * destruction must release engine-scoped borrowed storage.
 *
 * Run through `just ffi-c-harness`; the pre-fix process-wide channel cache
 * triggers an AddressSanitizer use-after-free when `first_output` is read
 * after the second engine replaces the shared cache entry.
 */

#include <stdint.h>
#include <stdio.h>
#include <string.h>

#include "ferric.h"

static FerricEngine *engine_with_output(const char *channel, const char *text) {
    char source[256] = {0};
    uint64_t fired = 0;
    FerricEngine *engine = ferric_engine_new();
    if (engine == NULL) {
        return NULL;
    }

    int written = snprintf(
        source, sizeof(source),
        "(defrule emit => (printout %s \"%s\"))", channel, text);
    if (written < 0 || (size_t)written >= sizeof(source) ||
        ferric_engine_load_string(engine, source) != FERRIC_ERROR_OK ||
        ferric_engine_reset(engine) != FERRIC_ERROR_OK ||
        ferric_engine_run(engine, -1, &fired) != FERRIC_ERROR_OK ||
        fired != UINT64_C(1)) {
        (void)ferric_engine_free(engine);
        return NULL;
    }

    return engine;
}

static int check_copy_contract(FerricEngine *engine) {
    char small[5] = {'x', 'x', 'x', 'x', 'x'};
    char exact[32] = {0};
    uintptr_t needed = 0;

    if (ferric_engine_get_output_copy(engine, "shared", NULL, 0, &needed) !=
            FERRIC_ERROR_OK ||
        needed != sizeof("engine-a")) {
        fprintf(stderr, "output copy size query returned the wrong result\n");
        return 1;
    }

    uintptr_t reported = 0;
    if (ferric_engine_get_output_copy(engine, "shared", small, sizeof(small),
                                      &reported) !=
            FERRIC_ERROR_BUFFER_TOO_SMALL ||
        reported != needed || small[sizeof(small) - 1] != '\0' ||
        strcmp(small, "engi") != 0) {
        fprintf(stderr, "undersized output copy was not reported explicitly\n");
        return 1;
    }

    if (ferric_engine_get_output_copy(engine, "shared", exact, needed,
                                      &reported) != FERRIC_ERROR_OK ||
        reported != needed || strcmp(exact, "engine-a") != 0) {
        fprintf(stderr, "exact output copy returned the wrong result\n");
        return 1;
    }

    reported = UINTPTR_MAX;
    if (ferric_engine_get_output_copy(engine, "missing", NULL, 0, &reported) !=
            FERRIC_ERROR_NOT_FOUND ||
        reported != 0) {
        fprintf(stderr, "missing output copy returned the wrong result\n");
        return 1;
    }

    return 0;
}

static int stress_engine_lifecycle(void) {
    for (unsigned i = 0; i < 128; i++) {
        char channel[32] = {0};
        int written =
            snprintf(channel, sizeof(channel), "lifecycle_%03u", i);
        if (written < 0 || (size_t)written >= sizeof(channel)) {
            return 1;
        }

        FerricEngine *engine = engine_with_output(channel, "payload");
        if (engine == NULL) {
            fprintf(stderr, "lifecycle engine %u failed to initialize\n", i);
            return 1;
        }
        const char *output = ferric_engine_get_output(engine, channel);
        if (output == NULL || strcmp(output, "payload") != 0 ||
            ferric_engine_free(engine) != FERRIC_ERROR_OK) {
            fprintf(stderr, "lifecycle engine %u failed\n", i);
            return 1;
        }
    }
    return 0;
}

int main(void) {
    FerricEngine *first = engine_with_output("shared", "engine-a");
    FerricEngine *second = engine_with_output("shared", "engine-b");
    if (first == NULL || second == NULL) {
        fprintf(stderr, "failed to create output-producing engines\n");
        if (first != NULL) {
            (void)ferric_engine_free(first);
        }
        if (second != NULL) {
            (void)ferric_engine_free(second);
        }
        return 1;
    }

    const char *first_output = ferric_engine_get_output(first, "shared");
    if (first_output == NULL || strcmp(first_output, "engine-a") != 0) {
        fprintf(stderr, "first engine returned the wrong output\n");
        return 1;
    }

    const char *second_output = ferric_engine_get_output(second, "shared");
    if (second_output == NULL || strcmp(second_output, "engine-b") != 0) {
        fprintf(stderr, "second engine returned the wrong output\n");
        return 1;
    }

    if (strcmp(first_output, "engine-a") != 0) {
        fprintf(stderr, "second engine invalidated first engine's pointer\n");
        return 1;
    }
    if (check_copy_contract(first) != 0) {
        return 1;
    }

    if (ferric_engine_free(first) != FERRIC_ERROR_OK ||
        ferric_engine_free(second) != FERRIC_ERROR_OK) {
        fprintf(stderr, "failed to free output-producing engines\n");
        return 1;
    }
    if (stress_engine_lifecycle() != 0) {
        return 1;
    }

    printf("output-lifetime harness: pointer isolation, copy, and cleanup passed\n");
    return 0;
}

File: crates/ferric-rules-ffi/src/tests/output_lifetime.rs
//! Regression coverage for engine-scoped borrowed output storage.

use crate::engine::{
    ferric_engine_clear_output, ferric_engine_free, ferric_engine_get_output,
    ferric_engine_get_output_copy, ferric_engine_last_error, ferric_engine_load_string,
    ferric_engine_new, ferric_engine_reset, ferric_engine_run, output_cache_entry_count_for_test,
    output_cache_lifetime_for_test, FerricEngine,
};
use crate::error::FerricError;
use std::ffi::{CStr, CString};

unsafe fn engine_with_output(text: &str) -> *mut FerricEngine {
    let engine = ferric_engine_new();
    assert!(!engine.is_null());
    let source = CString::new(format!(r#"(defrule emit => (printout shared "{text}"))"#)).unwrap();
    assert_eq!(
        ferric_engine_load_string(engine, source.as_ptr()),
        FerricError::Ok
    );
    assert_eq!(ferric_engine_reset(engine), FerricError::Ok);
    let mut fired = 0;
    assert_eq!(ferric_engine_run(engine, -1, &mut fired), FerricError::Ok);
    assert_eq!(fired, 1);
    engine
}

#[test]
fn borrowed_output_cache_is_engine_scoped_and_freed_with_engine() {
    unsafe {
        let channel = CString::new("shared").unwrap();
        let first = engine_with_output("engine-a");
        let second = engine_with_output("engine-b");

        assert!(!ferric_engine_get_output(first, channel.as_ptr()).is_null());
        assert_eq!(output_cache_entry_count_for_test(first), Some(1));
        let first_entry = output_cache_lifetime_for_test(first, "shared")
            .expect("first engine should have a cached output entry");

        assert!(!ferric_engine_get_output(second, channel.as_ptr()).is_null());
        assert_eq!(output_cache_entry_count_for_test(first), Some(1));
        assert_eq!(output_cache_entry_count_for_test(second), Some(1));
        let first_entry_survived_second_read = first_entry.upgrade().is_some();
        let second_entry = output_cache_lifetime_for_test(second, "shared")
            .expect("second engine should have a cached output entry");

        assert_eq!(ferric_engine_free(first), FerricError::Ok);
        assert_eq!(ferric_engine_free(second), FerricError::Ok);
        let first_entry_was_reclaimed = first_entry.upgrade().is_none();
        let second_entry_was_reclaimed = second_entry.upgrade().is_none();

        assert!(
            first_entry_survived_second_read,
            "reading engine B must not replace engine A's same-channel cache entry"
        );
        assert!(
            first_entry_was_reclaimed && second_entry_was_reclaimed,
            "freeing each engine must reclaim its cached output entries"
        );
    }
}

#[test]
fn clearing_output_and_reset_reclaim_borrowed_snapshots() {
    unsafe {
        let engine = engine_with_output("engine-a");
        let channel = CString::new("shared").unwrap();
        assert!(!ferric_engine_get_output(engine, channel.as_ptr()).is_null());
        assert_eq!(output_cache_entry_count_for_test(engine), Some(1));
        let cleared_entry = output_cache_lifetime_for_test(engine, "shared").unwrap();

        assert_eq!(
            ferric_engine_clear_output(engine, channel.as_ptr()),
            FerricError::Ok
        );
        assert_eq!(output_cache_entry_count_for_test(engine), Some(0));
        assert!(cleared_entry.upgrade().is_none());

        assert_eq!(ferric_engine_reset(engine), FerricError::Ok);
        let mut fired = 0;
        assert_eq!(ferric_engine_run(engine, -1, &mut fired), FerricError::Ok);
        assert_eq!(fired, 1);
        assert!(!ferric_engine_get_output(engine, channel.as_ptr()).is_null());
        let reset_entry = output_cache_lifetime_for_test(engine, "shared").unwrap();

        assert_eq!(ferric_engine_reset(engine), FerricError::Ok);
        assert_eq!(output_cache_entry_count_for_test(engine), Some(0));
        assert!(reset_entry.upgrade().is_none());
        assert_eq!(ferric_engine_free(engine), FerricError::Ok);
    }
}

#[test]
fn output_copy_reports_required_length_and_truncation() {
    unsafe {
        let engine = engine_with_output("engine-a");
        let channel = CString::new("shared").unwrap();
        let mut needed = 0;

        assert_eq!(
            ferric_engine_get_output_copy(
                engine,
                channel.as_ptr(),
                std::ptr::null_mut(),
                0,
                &mut needed,
            ),
            FerricError::Ok
        );
        assert_eq!(needed, "engine-a".len() + 1);

        let sentinel = std::os::raw::c_char::try_from(b'x')
            .expect("ASCII sentinel must fit in the platform C char type");
        let mut small = [sentinel; 5];
        let mut reported = 0;
        assert_eq!(
            ferric_engine_get_output_copy(
                engine,
                channel.as_ptr(),
                small.as_mut_ptr(),
                small.len(),
                &mut reported,
            ),
            FerricError::BufferTooSmall
        );
        assert_eq!(reported, needed);
        assert_eq!(
            CStr::from_ptr(small.as_ptr()).to_bytes(),
            b"engi",
            "undersized copies must be explicitly reported and NUL-terminated"
        );

        let mut exact = vec![0 as std::os::raw::c_char; needed];
        assert_eq!(
            ferric_engine_get_output_copy(
                engine,
                channel.as_ptr(),
                exact.as_mut_ptr(),
                exact.len(),
                &mut reported,
            ),
            FerricError::Ok
        );
        assert_eq!(reported, needed);
        assert_eq!(CStr::from_ptr(exact.as_ptr()).to_bytes(), b"engine-a");

        assert_eq!(ferric_engine_free(engine), FerricError::Ok);
    }
}

#[test]
fn output_copy_validates_missing_and_invalid_inputs() {
    unsafe {
        let engine = engine_with_output("engine-a");
        let channel = CString::new("shared").unwrap();
        let missing = CString::new("missing").unwrap();
        let mut out_len = usize::MAX;
        let mut byte = 0 as std::os::raw::c_char;

        assert_eq!(
            ferric_engine_get_output_copy(
                engine,
                missing.as_ptr(),
                std::ptr::null_mut(),
                0,
                &mut out_len,
            ),
            FerricError::NotFound
        );
        assert_eq!(out_len, 0);
        assert_eq!(
            CStr::from_ptr(ferric_engine_last_error(engine))
                .to_str()
                .unwrap(),
            "output channel has no captured output: missing"
        );

        out_len = usize::MAX;
        assert_eq!(
            ferric_engine_get_output_copy(
                engine,
                std::ptr::null(),
                std::ptr::null_mut(),
                0,
                &mut out_len,
            ),
            FerricError::NullPointer
        );
        assert_eq!(out_len, 0);

        assert_eq!(
            ferric_engine_get_output_copy(
                engine,
                channel.as_ptr(),
                std::ptr::null_mut(),
                1,
                &mut out_len,
            ),
            FerricError::InvalidArgument
        );
        assert_eq!(out_len, "engine-a".len() + 1);

        assert_eq!(
            ferric_engine_get_output_copy(engine, channel.as_ptr(), &mut byte, 0, &mut out_len,),
            FerricError::BufferTooSmall
        );
        assert_eq!(out_len, "engine-a".len() + 1);

        assert_eq!(
            ferric_engine_get_output_copy(
                engine,
                channel.as_ptr(),
                std::ptr::null_mut(),
                0,
                std::ptr::null_mut(),
            ),
            FerricError::InvalidArgument
        );

        out_len = usize::MAX;
        assert_eq!(
            ferric_engine_get_output_copy(
                std::ptr::null(),
                channel.as_ptr(),
                std::ptr::null_mut(),
                0,
                &mut out_len,
            ),
            FerricError::NullPointer
        );
        assert_eq!(out_len, 0);

        assert_eq!(ferric_engine_free(engine), FerricError::Ok);
    }
}
